### PR TITLE
[TUTORIAL][ANSOR] Using the template-free auto-scheduler on CPU

### PR DESCRIPTION
**Pull request by merrymercy**
Add a tutorial on how to auto-schedule a subgraph for CPU.

Any feedback on the API is welcomed.

cc @jcf94 @tqchen @junrushao1994 @comaniac @jwfromm

### CODE REVIEW

File: python/tvm/auto_scheduler/auto_schedule.py [hunk 1]
```diff
@@ -156,16 +159,41 @@ def __init__(
         )
 
 
+def create_task(func, args, target, target_host=None, hardware_params=None):
```
Comment (jcf94) [inline python/tvm/auto_scheduler/auto_schedule.py:162]: I'm thinking about what's the different of calling `SearchTask()` directly, seems they just have different input parameters.

I notice that AutoTVM used an api of `autotvm.task.create()`, should we keep them the same?

**Review by jcf94 — COMMENTED** [review]
Fine to me.

**Review by jcf94 — APPROVED** [review]
Seems we can next migrate the tuning scripts here, and start to reproduce the op/subgraph performance in our paper based on the new master.

**Review by junrushao1994 — APPROVED** [review]
LGTM

File: python/tvm/auto_scheduler/auto_schedule.py (hunk 1 above)
Comment (merrymercy) [reply python/tvm/auto_scheduler/auto_schedule.py:162]: Python does not support constructor overloading, so I have to create a new function for this.

I think both `auto_scheduler.create_task` and `auto_scheduler.task.create` are okay.

File: tutorials/auto_scheduler/tune_matmul_x86.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    matmul = te.compute((N, M), lambda i, j: te.sum(A[i, k] * B[k, j], axis=k), name="matmul")
+    out = te.compute((N, M), lambda i, j: matmul[i, j] + C[i, j], name="out")
+
+    return [A, B, C, out]
+
+
+######################################################################
+# Create the search task
+# ^^^^^^^^^^^^^^^^^^^^^^
+# We then create the a search task with N=L=M=128 and dtype="float32"
```
Comment (tmoreau89) [inline tutorials/auto_scheduler/tune_matmul_x86.py:61]: remove the/a

**Review by tqchen — APPROVED** [review]

File: tutorials/auto_scheduler/tune_matmul_x86.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+tune_option = auto_scheduler.TuningOptions(
+    num_measure_trials=10, measure_callbacks=[auto_scheduler.RecordToFile("matmul.json")]
+)
+
+######################################################################
+# Run the search
+# ^^^^^^^^^^^^^^
+# Now we get all inputs ready. Pretty simple, isn't it?
+# We can kick off the search and let the auto-scheduler do its magic.
+# After some measurement trials, it will return the best schedule it founds.
```
Comment (tmoreau89) [inline tutorials/auto_scheduler/tune_matmul_x86.py:89]: it found

File: tutorials/auto_scheduler/tune_matmul_x86.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+sch, args = auto_scheduler.auto_schedule(task, tuning_options=tune_option)
+
+######################################################################
+# We can lower schedule to see the IR after auto-scheduling.
```
Comment (tmoreau89) [inline tutorials/auto_scheduler/tune_matmul_x86.py:94]: lower the schedule

File: tutorials/auto_scheduler/tune_matmul_x86.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+func(tvm.nd.array(a_np), tvm.nd.array(b_np), tvm.nd.array(c_np), d_tvm)
+
+tvm.testing.assert_allclose(d_np, d_tvm.asnumpy(), rtol=1e-3)
+
+######################################################################
+# Using the record file
+# ^^^^^^^^^^^^^^^^^^^^^
+# During the search, all measuremnt records are dumpped into the record
+# file "matmul.json". The measurement records can be used to resume the
+# search, re-apply search results and other analysis.
```
Comment (tmoreau89) [inline tutorials/auto_scheduler/tune_matmul_x86.py:121]: and perform other analyses.

File: tutorials/auto_scheduler/tune_matmul_x86.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+#
+# Here we show an example where we load the best schedule from a file,
+# print the equivalent python schedule API, and build the binary again.
+
+# Load the measuremnt record for the best schedule
+inp, res = auto_scheduler.load_best("matmul.json", task.workload_key)
+
+# Print equivalent python schedule API. This can be used for debugging and
+# learning the behavior of auto-scheduler.
```
Comment (tmoreau89) [inline tutorials/auto_scheduler/tune_matmul_x86.py:130]: "of the autoscheduler"

**Review by tmoreau89 — COMMENTED** [review]
Nice work! I've only left cosmetic comments.

File: python/tvm/auto_scheduler/auto_schedule.py [hunk 7]
```diff
@@ -156,16 +159,41 @@ def __init__(
...
+    """Create a search task
+
+    Parameters
+    ----------
+    func : Union[Function, str]
+        The function that returns the compute declaration Tensors.
+        Can be the a function or the function name.
+    args : Args
```
Comment (comaniac) [inline python/tvm/auto_scheduler/auto_schedule.py:170]:
```suggestion
    args : Union[Tuple[Any, ...], List[Any]]
```

File: python/tvm/auto_scheduler/auto_schedule.py [hunk 8]
```diff
@@ -156,16 +159,41 @@ def __init__(
...
+    target : tvm.target.Target
+        The target device of this search task.
+    target_host : Optional[tvm.target.Target]
+        The target host device of this search task.
+    hardware_params : Optional[HardwareParams]
+        Hardware parameters used in this search task.
+
+    Returns
+    -------
+        task : the created task
```
Comment (comaniac) [inline python/tvm/auto_scheduler/auto_schedule.py:181]:
```suggestion
        SearchTask: the created task
```

File: tutorials/auto_scheduler/tune_matmul_x86.py (hunk 5 above)
Comment (comaniac) [inline tutorials/auto_scheduler/tune_matmul_x86.py:121]: It would be better to also show how to resume the search with existing logs.

**Review by comaniac — APPROVED** [review]
LGTM. Just nits.

**Review by tmoreau89 — APPROVED** [review]
Thanks @merrymercy LGTM

File: tutorials/auto_scheduler/tune_matmul_x86.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Using the template-free auto-scheduler on CPU 
+=============================================
+**Author**: `Lianmin Zheng <https://github.com/merrymercy>`_, \
+            `Chengfan Jia <https://github.com/jcf94/>`_
+
+Different from the exiting :ref:`autotvm <tutorials-autotvm-sec>` which relies on
```
Comment (masahi) [inline tutorials/auto_scheduler/tune_matmul_x86.py:23]: existing

**Review by mbaret — APPROVED** [review]
I've just had a quick play around with this tutorial and it looks good to me. There is one thing I wonder though, which is that even when I boost the number of trials it never seems to decide to inline the bias add (x86). Is this a limitation of Ansor or is inlining just not performant here?

**Comment by merrymercy** [comment]
@mbaret Good catch! It is actually a small bug introduced during the upstream.
Ansor is designed to work with arbitrary subgraph and it should fuse (inline) them correctly.

File: tutorials/auto_scheduler/tune_matmul_x86.py (hunk 5 above)
Comment (merrymercy) [reply tutorials/auto_scheduler/tune_matmul_x86.py]: The example is added

**Comment by mbaret** [comment]
Ah nice, it's working as I expected now.

File: tutorials/auto_scheduler/tune_matmul_x86.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+)
+
+######################################################################
+# Run the search
+# ^^^^^^^^^^^^^^
+# Now we get all inputs ready. Pretty simple, isn't it?
+# We can kick off the search and let the auto-scheduler do its magic.
+# After some measurement trials, it will return the best schedule it found.
+
+sch, args = auto_scheduler.auto_schedule(task, tuning_options=tune_option)
```
Comment (binarybana) [inline tutorials/auto_scheduler/tune_matmul_x86.py:91]: It'd be nice to briefly describe what the return values of `sch` and `args` _are_ here. I'm assuming that `sch` is a schedule, but is it encoded in TIR, or some other data structure? Same with `args`.

Comment (jcf94) [reply tutorials/auto_scheduler/tune_matmul_x86.py:91]: If you check the doc string of the `auto_scheduler.auto_schedule()`, you'll find the explaination of the return values: A `te.schedule` and the a list of `te.Tensor` to be used in `tvm.lower` or `tvm.build`.

Comment (merrymercy) [reply tutorials/auto_scheduler/tune_matmul_x86.py:91]: I will merge this and test the docstring link on the tvm website.
If it is not working, I can add more explanations in the next PR, which is a similar tutorial for GPU.

**Pull request merged**